File: 2015/day_1/main.c
#include <stdio.h>
#include <string.h>
#include <time.h>

int main() {
    // Open Filepointer
    FILE *fptr = fopen("input.txt", "r");

    // if file not found, pointer is null
    if (!fptr) {
        printf("No File found!");
        return 1;
    }

    clock_t start = clock();

    int floorCount = 0;
    char c;

    int index = 1;
    // standard C has no bool, so I abuse the enteredBasement var to also store the value
    int enteredBasement = 0;

    // check file, char by char
    do {
        c = fgetc(fptr);

        if (c == '(') {
            floorCount ++;
        } else if (c == ')') {
            floorCount --;
            if (floorCount < 0 && !enteredBasement) {
                enteredBasement = index;
            }
        }   
        index ++;
    }
    while (c != EOF);

    clock_t end = clock();

    fclose(fptr);

    printf("Santa got up to the %d floor\n", floorCount);
    printf("Santa entered the basement first at instruction %d", enteredBasement);
    float seconds = (float)(end - start) / CLOCKS_PER_SEC;
    printf("\nNeeded Time: %f seconds", seconds);

    return 0;
}

File: 2015/day_1/main.py
import time

def find_floor():
    with open("input.txt", "r") as f:
        floor_instructions = f.read()

    start = time.time()

    floor_count = 0
    entered_basement = False

    # linear time
    for i, instruction in enumerate(floor_instructions):
        if instruction == "(":
            floor_count += 1
        elif instruction == ")":
            floor_count -= 1

        if floor_count < 0 and not entered_basement:
            basement_index = i
            entered_basement = True

    end = time.time()

    print(f"Santa entered the basement first at instruction {basement_index + 1 }")
    print(f"Santa is taken to floor {floor_count}")
    print(f"Needed {end - start} seconds")

if __name__ == "__main__":
    find_floor()